### PR TITLE
Form::getMessages() on plain array collection - fix for issue #135

### PR DESCRIPTION
**Pull request by vaclavvanik**
I used this [fix](https://github.com/zendframework/zendframework/issues/5265#issuecomment-56173075). If anyone has better idea let me know!

### CODE REVIEW

File: test/FieldsetTest.php [hunk 1]
```diff
@@ -572,4 +572,19 @@ public function testBindValuesPreservesNewValueAfterValidation()
 
         $this->assertSame('New value', $form->get('foo')->getValue());
     }
+
+    /**
+     * Error test for https://github.com/zendframework/zend-form/issues/135
+     */
+    public function testRetrieveErrorMessagesForArrayInputCollection()
+    {
+        $form = new TestAsset\ArrayInputCollectionForm();
```
Comment (froschdesign) [inline test/FieldsetTest.php]: Please test here the `Fieldset` not the `Form` class or the `Collection`element.

File: test/FieldsetTest.php [hunk 2]
```diff
@@ -572,4 +572,19 @@ public function testBindValuesPreservesNewValueAfterValidation()
...
+        $form->setData([]);
+        $form->isValid();
+
+        $messages = $form->getMessages();
+
+        $this->assertArrayHasKey('foo', $messages);
+        $this->assertArrayNotHasKey('bar', $messages);
```
Comment (froschdesign) [inline test/FieldsetTest.php]: What is `bar` and why do you test it?

File: test/TestAsset/ArrayInputCollectionForm.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace ZendTest\Form\TestAsset;
+
+use Zend\Form;
+use Zend\InputFilter\ArrayInput;
+use Zend\InputFilter\InputFilterProviderInterface;
+
+class ArrayInputCollectionForm extends Form\Form implements InputFilterProviderInterface
```
Comment (froschdesign) [inline test/TestAsset/ArrayInputCollectionForm.php]: I think, the class is not needed. Please have a look at [`ZendTest\Form\Element\CollectionTest`](https://github.com/zendframework/zend-form/blob/master/test/Element/CollectionTest.php).

**Comment by vaclavvanik** [comment]
@froschdesign I think about the test again and again. Basically this patch allows storing messages for elements which are not founds. It does not break any other test. I would rewrite test to:

```
    public function testSetAndGetErrorMessagesForNonExistentElements()
    {
        $messages = [
            'foo' => [
                'foo_message_key' => 'foo_message_val',
            ],
            'bar' => [
                'bar_message_key' => 'bar_message_val',
            ],
        ];

        $fieldset = new Fieldset();
        $fieldset->setMessages($messages);

        $this->assertEquals($messages, $fieldset->getMessages());
    }
```

agreed?

**Comment by froschdesign** [comment]
This works:

```php
// ZendTest\Form\Element\CollectionTest

public function testGetErrorMessagesForInvalidCollectionElements()
{
    // Configure InputFilter
    $inputFilter = $this->form->getInputFilter();
    $inputFilter->add(
        [
            'name'     => 'colors',
            'type'     => ArrayInput::class,
            'required' => true,
        ]
    );
    $inputFilter->add(
        [
            'name'     => 'fieldsets',
            'type'     => ArrayInput::class,
            'required' => true,
        ]
    );


    $this->form->setData([]);
    $this->form->isValid();

    $this->assertEquals(
        [
            'colors'     => [
                'isEmpty' => "Value is required and can't be empty",
            ],
            'fieldsets' => [
                'isEmpty' => "Value is required and can't be empty",
            ],
        ],
        $this->form->getMessages()
    );
}
```

And this fails:
```php
// ZendTest\Form\Element\CollectionTest

public function testValidationForInvalidChildElements()
{
    // Configure InputFilter
    $inputFilter = $this->form->getInputFilter();
    $inputFilter->add(
        [
            'name'     => 'colors',
            'type'     => ArrayInput::class,
            'required' => true,
        ]
    );
    $inputFilter->add(
        [
            'name'     => 'fieldsets',
            'type'     => ArrayInput::class,
            'required' => true,
        ]
    );

    $this->form->setData(
        [
            'colors'    => [],
            'fieldsets' => [],
        ]
    );

    $this->assertFalse($this->form->isValid());
}
```

**Comment by vaclavvanik** [comment]
First look at `testValidationForInvalidChildElements()` => its problem in InputFilter.

Code from `Form::isValid()`

```
        $this->isValid = $result = $filter->isValid();
        ...
        if (!$result) {
            $this->setMessages($filter->getMessages());
        }
```

Error messages are set only when validation fails. This test is imho unrelated to my patch.

**Comment by froschdesign** [comment]
@vaclavvanik 
> First look at testValidationForInvalidChildElements() => its problem in InputFilter.

Sorry, copy&paste error! I've updated my previous comment.

**Comment by froschdesign** [comment]
@vaclavvanik 
👍

**Comment by vaclavvanik** [comment]
@froschdesign tests updated. I hope I did not miss anything.

**Comment by vaclavvanik** [comment]
@froschdesign can you merge if don't have any other ideas?

**Comment by akrabat** [comment]
Why is this change on the Fieldset class and not the Collection class?

**Comment by froschdesign** [comment]
@akrabat 
The `Fieldset` class resets all previous added messages. Look at this unit test: https://github.com/vaclavvanik/zend-form/blob/b5c442dbab50d9817e3c965aaf083c6b6bf3fed0/test/FieldsetTest.php#L579

**Comment by akrabat** [comment]
I mean, why isn't this fixed as new methods `setMessages` and `getMessages` on `Collection`? 

i.e. does this bug affect `Fieldset` as well as `Collection`?

**Comment by froschdesign** [comment]
> i.e. does this bug affect Fieldset as well as Collection?

Right!

**Comment by akrabat** [comment]
> > i.e. does this bug affect Fieldset as well as Collection?
> Right!

That's a "yes it does" I assume. Next question. Will this have any side-effects for people using normal elements inside `Fieldset`s?

**Comment by froschdesign** [comment]
@akrabat 
Sorry, for the late response.

> Will this have any side-effects for people using normal elements inside Fieldsets?

No and the [unit tests for these scenario](https://github.com/zendframework/zend-form/blob/master/test/FieldsetTest.php#L242)s says the same.

**Comment by weierophinney** [comment]
Thanks, @vaclavvanik

**Pull request merged**